File: src/ChannelControl.cpp
#include "ChannelControl.h"
#include "StdAfx.h"

CChannelControl::CChannelControl(unsigned int channelCount) : m_channelCount(channelCount) {

    for (unsigned int i = 0; i < channelCount; i++) {
        m_channelon.push_back(false);
    }
}


bool CChannelControl::IsChannelOn(const ChannelNumber channel) {
    return m_channelon[channel];
}


// ----------------------------------------------------------------------------
// Channel On/Off helper functions

/// <summary>
/// Turn all or individual sound channels on or off
/// </summary>
/// <param name="ch">-1 = all channels, 0 - 7 = the sound channel</param>
/// <param name="onoff">-1 = invert state, 0 = off, 1 = on</param>
void CChannelControl::SetChannelOnOff(const int ch, int onoff)
{
    if (ch < 0)
    {
        // All channels
        if (onoff >= 0)
            for (unsigned int i = 0; i < m_channelon.size(); i++) { m_channelon[i] = onoff; } // Set the given on/off state
        else
            for (unsigned int i = 0; i < m_channelon.size(); i++) { m_channelon[i] = !m_channelon[i]; } // Invert the on/off state
    }
    else if (ch < m_channelon.size())
    {
        // Just that one
        if (onoff >= 0) {
            m_channelon[ch] = onoff;	// Set the given on/off state
        }
        else {
            m_channelon[ch] = !m_channelon[ch];		// Invert the on/off state
        }
    }
}

void CChannelControl::ToggleChannelOnOff(const ChannelNumber channelNumber) {
    SetChannelOnOff(channelNumber, -1);
}

void CChannelControl::SetAllChannelsOff() {
    SetChannelOnOff(-1, 0);
}

void CChannelControl::SetAllChannelsOn() {
    SetChannelOnOff(-1, 1);
}

void CChannelControl::ToggleAllChannelsOnOff() {
    SetChannelOnOff(-1, -1);
}

/// <summary>
/// Turn ON only one channel, making sure that all others are turned off
/// </summary>
/// <param name="ch"></param>
void CChannelControl::SetChannelSolo(const ChannelNumber channelNumber)
{
    if (IsChannelOn(channelNumber))
    {
        // Target channel is ON
        // If any other channel is on then turn them all off except for the target channel
        for (unsigned int i = 0; i < m_channelCount; i++)
        {
            if (i != channelNumber && IsChannelOn(i)) goto Channel_SOLO;
        }
        // All other channels are off, turn them all on
        SetAllChannelsOn();
    }
    else
    {
        // Target channel is OFF
    Channel_SOLO:
        SetAllChannelsOff();
        SetChannelOnOff(channelNumber, 1);	// and turn ON only the solo channel
    }

}

File: src/java/com/wudsn/tools/rmt/model/ChannelControl.java
package com.wudsn.tools.rmt.model;

import java.util.Arrays;

/**
 * Ported from CChannelControl (src/cpp/ChannelControl.h/.cpp).
 *
 * <p>C++'s private SetChannelOnOff(ch, onoff) multiplexes "one channel or
 * all channels" (ch == -1) and "set or toggle" (onoff == -1) behind two
 * sentinel parameters, then uses a {@code goto} inside setChannelSolo() to
 * share its "turn everything off, then turn just the target on" tail with
 * the "target is off" case. Ported here as direct, single-purpose methods
 * instead - Java has no goto, and the sentinel-multiplexing added no value
 * once each call site can just call the specific method it means; this
 * doesn't change any observable behavior (verified by tracing
 * setChannelSolo()'s two branches against the original goto/sentinel
 * version before simplifying).
 */
public final class ChannelControl {

	private final boolean[] channelOn;

	public ChannelControl(int channelCount) {
		channelOn = new boolean[channelCount];
	}

	public boolean isChannelOn(int channelNumber) {
		return channelOn[channelNumber];
	}

	public void toggleChannelOnOff(int channelNumber) {
		channelOn[channelNumber] = !channelOn[channelNumber];
	}

	public void setAllChannelsOn() {
		Arrays.fill(channelOn, true);
	}

	public void setAllChannelsOff() {
		Arrays.fill(channelOn, false);
	}

	public void toggleAllChannelsOnOff() {
		for (int i = 0; i < channelOn.length; i++) {
			channelOn[i] = !channelOn[i];
		}
	}

	/**
	 * Turn ON only one channel, making sure all others are turned off. Soloing an already-solo'd channel (the only one on) turns everything back on instead of leaving it soloed.
	 */
	public void setChannelSolo(int channelNumber) {
		boolean anyOtherChannelOn = false;
		for (int i = 0; i < channelOn.length; i++) {
			if (i != channelNumber && channelOn[i]) {
				anyOtherChannelOn = true;
				break;
			}
		}
		if (isChannelOn(channelNumber) && !anyOtherChannelOn) {
			setAllChannelsOn();
		} else {
			setAllChannelsOff();
			channelOn[channelNumber] = true;
		}
	}
}
